4.x - Form Context

ADmad opened the pull request:
I moved the generation of default html attributes using context from `FormHelper` to widgets wherever possible.

I will update method docblocks if the changes seem reasonable.

ADmad committed:
Move introspection of default value for "step" into widget

ADmad committed:
Cleanup unneeded values.

ADmad committed:
Move setting default maxlength into widgets

ADmad committed:
Remove unneeded code

ADmad commented:
Currently if not explicitly specified values for attributes like `maxlength` are only set if using `FormHelper::contro()` and not when using individual input methods like `FormHelper::text()/textarea()`.

Doesn't is make sense to always set them whenever context is available?

dereuromark commented:
> Doesn't it make sense to always set them whenever context is available?

I would totally say yes. Usually the main reason to use specific ones instead of control() is to remove the diff (since this is not possible that easy anymore in 3.x now). The whole form input part should be the same IMO.

ADmad committed:
Set "fieldName" in widget data even for individual input methods.

ADmad committed:
Reduce code duplication by making BasicWidget easily extendable.

markstory commented on src/View/Form/EntityContext.php:
```diff
@@ -538,15 +538,20 @@ public function getMaxLength(string $field): ?int
         }
 
+        $attributes = $this->attributes($field);
+        if (!empty($attributes['length'])) {
```
If the length is 0 should we return it?

ADmad replied:
I don't think db columns with length `0` or `maxlength` attribute with `'0'` has any practical use case, so such a default doesn't make much sense.

markstory commented on src/View/Widget/BasicWidget.php:
```diff
@@ -72,6 +74,24 @@ public function render(array $data, ContextInterface $context): string
+            && !isset($data['step'])
+            && $fieldName
+        ) {
+            $data = $this->setStep($data, $context, $data['fieldName']);
```
Should this be `$fieldName` instead of `$data['fieldName']`?

ADmad replied:
Yup, done.

markstory commented on src/View/Widget/HtmlAttributesTrait.php:
```diff
@@ -0,0 +1,32 @@
+
+use Cake\View\Form\ContextInterface;
+
+trait HtmlAttributesTrait
```
Could we mark this trait as an internal one with a doc string?

ADmad replied:
Done

ADmad committed:
Fix "required" attribute being set for hidden fields.

ADmad committed:
Allow ContextInterface::isRequired() to return null.

The method now returns boolean only if it can deterministically say
whether a field is allowed to be empty or not.

This change allows FormHelper::select() to set it's `empty` option to `true` by default
is the field is explicitly allowed to be empty.

ADmad committed:
Use variable

ADmad committed:
Mark HtmlAttributesTrait as internal.

ADmad commented:
- Default HTML attributes based on context introspection are now set when directly using input specific methods too.
- `FormHelper::select()` now sets `empty` option to `true` by default if the field has been explicitly allowed to be empty in validator. Earlier it always defaulted to `'empty' => false` regardless of whether the field was allowed to be empty or not. 
This required changed required allowing `ContextInterface::isRequired()` return nullable, so that `EntityContext::isRequired()` returns `null` if no validations are set for the field. This way the `empty` option still defaults to `false` is no `notEmpty()` or `allowEmpty()` is not explicity set for the field.

ADmad committed:
Update docblocks

ADmad committed:
Ignore invalid phptan errors

ADmad commented:
The failure on travis for PHP 7.4 and Mysql seems related to the changes https://travis-ci.org/cakephp/cakephp/jobs/611047166#L577

But there are no failures for other PHP versions and mysql on GH so not sure what's causing the failure on travis.

ADmad committed:
Fix test

ADmad committed:
Remove duplicated call

ADmad committed:
Merge branch '4.x' into 4.x-form-context

ADmad committed:
Move step setting for datetime fields into HtmlAttributesTrait::setStep()

othercorey commented on src/View/Widget/DateTimeWidget.php:
```diff
@@ -110,25 +109,15 @@ public function render(array $data, ContextInterface $context): string
 
         if (!isset($data['step'])) {
-            $step = $this->defaultStep[$data['type']];
+            $data['step'] = $this->defaultStep[$data['type']];
```
Shouldn't this default step go in the magic setStep() function too?

ADmad replied:
The `setStep()` has `$fieldName` as an argument so I called it only if `$data['fieldName']` is set.

Also I got rid of the trait since put it methods directly in `BasicWidget` since all other widgets extend it anyway. Also `setStep()` is now overridden in `DateTimeWidget()` so introspection related to datetime fields is contained within same class.

othercorey replied:
It looks good the way it is now.  I approved.

ADmad committed:
Move HtmlAttributesTrait methods into BasicWiget.

Given that all other widgets extends BasicWidget having the methods
in a trait was unnecessary.

ADmad committed:
Move setting of step for date time inputs into DateTimeWidget

ADmad commented on src/View/Form/ArrayContext.php:
```diff
@@ -199,11 +199,25 @@ public function val(string $field, array $options = [])
     {
-        return (bool)$this->getRequiredMessage($field);
+        if (!is_array($this->_context['required'])) {
+            return null;
```
I have explained in 2nd point of my earlier comment https://github.com/cakephp/cakephp/pull/13875#issuecomment-553090729.
